File: packages/sdk/java/src/main/java/com/ihui/ai/sdk/module/ToolsApi.java
package com.ihui.ai.sdk.module;

import com.fasterxml.jackson.databind.JsonNode;
import com.ihui.ai.sdk.BaseClient;
import com.ihui.ai.sdk.model.ToolCallRequest;

import java.util.Map;

/**
 * MCP 工具 / 技能 / 人格 / 代码搜索 / 截图模块。
 *
 * <p>端点(16 个):
 * <ul>
 *   <li>GET  /v1/tools</li>
 *   <li>POST /v1/tools/call</li>
 *   <li>GET  /v1/resources</li>
 *   <li>GET  /v1/resources/:uri</li>
 *   <li>GET  /v1/prompts</li>
 *   <li>POST /v1/prompts/invoke</li>
 *   <li>GET  /v1/skills</li>
 *   <li>GET  /v1/slash-commands</li>
 *   <li>POST /v1/slash-commands</li>
 *   <li>POST /v1/sampling</li>
 *   <li>GET  /v1/personas</li>
 *   <li>GET  /v1/personas/:name</li>
 *   <li>POST /v1/tools/search-codebase</li>
 *   <li>POST /v1/tools/search-web</li>
 *   <li>POST /v1/tools/analyze-code</li>
 *   <li>POST /v1/screenshot</li>
 * </ul>
 */
public final class ToolsApi {

    private final BaseClient client;

    /**
     * 构造 ToolsApi。
     *
     * @param client 底层 BaseClient
     */
    public ToolsApi(BaseClient client) {
        this.client = client;
    }

    /**
     * GET /v1/tools(MCP 工具列表)。
     *
     * @return 工具列表(JsonNode)
     */
    public JsonNode list() {
        return client.request("GET", "/tools", null, JsonNode.class);
    }

    /**
     * POST /v1/tools/call(调用 MCP 工具)。
     *
     * @param req 工具调用请求
     * @return 调用结果(JsonNode)
     */
    public JsonNode call(ToolCallRequest req) {
        return client.request("POST", "/tools/call", req, JsonNode.class);
    }

    /**
     * GET /v1/resources(MCP 资源列表)。
     *
     * @return 资源列表(JsonNode)
     */
    public JsonNode listResources() {
        return client.request("GET", "/resources", null, JsonNode.class);
    }

    /**
     * GET /v1/resources/:uri(资源详情)。
     *
     * @param uri 资源 URI
     * @return 资源详情(JsonNode)
     */
    public JsonNode getResource(String uri) {
        return client.request("GET", "/resources/" + BaseClient.encode(uri),
                null, JsonNode.class);
    }

    /**
     * GET /v1/prompts(MCP 提示词列表)。
     *
     * @return 提示词列表(JsonNode)
     */
    public JsonNode listPrompts() {
        return client.request("GET", "/prompts", null, JsonNode.class);
    }

    /**
     * POST /v1/prompts/invoke(调用提示词)。
     *
     * @param req 请求体(name / arguments 等)
     * @return 调用结果(JsonNode)
     */
    public JsonNode invokePrompt(Map<String, Object> req) {
        return client.request("POST", "/prompts/invoke", req, JsonNode.class);
    }

    /**
     * GET /v1/skills(技能列表)。
     *
     * @return 技能列表(JsonNode)
     */
    public JsonNode listSkills() {
        return client.request("GET", "/skills", null, JsonNode.class);
    }

    /**
     * GET /v1/slash-commands(slash 命令列表)。
     *
     * @return 命令列表(JsonNode)
     */
    public JsonNode listSlashCommands() {
        return client.request("GET", "/slash-commands", null, JsonNode.class);
    }

    /**
     * POST /v1/slash-commands(调用 slash 命令)。
     *
     * @param req 请求体(command / args)
     * @return 调用结果(JsonNode)
     */
    public JsonNode invokeSlashCommand(Map<String, Object> req) {
        return client.request("POST", "/slash-commands", req, JsonNode.class);
    }

    /**
     * POST /v1/sampling(模型采样)。
     *
     * @param req 请求体(messages / model 等)
     * @return 采样结果(JsonNode)
     */
    public JsonNode sampling(Map<String, Object> req) {
        return client.request("POST", "/sampling", req, JsonNode.class);
    }

    /**
     * GET /v1/personas(人格列表)。
     *
     * @return 人格列表(JsonNode)
     */
    public JsonNode listPersonas() {
        return client.request("GET", "/personas", null, JsonNode.class);
    }

    /**
     * GET /v1/personas/:name(人格详情)。
     *
     * @param name 人格名称
     * @return 人格详情(JsonNode)
     */
    public JsonNode getPersona(String name) {
        return client.request("GET", "/personas/" + BaseClient.encode(name),
                null, JsonNode.class);
    }

    /**
     * POST /v1/tools/search-codebase(代码库搜索)。
     *
     * @param req 请求体(query / filePattern 等)
     * @return 搜索结果(JsonNode)
     */
    public JsonNode searchCodebase(Map<String, Object> req) {
        return client.request("POST", "/tools/search-codebase", req, JsonNode.class);
    }

    /**
     * POST /v1/tools/search-web(网页搜索)。
     *
     * @param req 请求体(query / count 等)
     * @return 搜索结果(JsonNode)
     */
    public JsonNode searchWeb(Map<String, Object> req) {
        return client.request("POST", "/tools/search-web", req, JsonNode.class);
    }

    /**
     * POST /v1/tools/analyze-code(代码分析)。
     *
     * @param req 请求体(code / language / analysis 等)
     * @return 分析结果(JsonNode)
     */
    public JsonNode analyzeCode(Map<String, Object> req) {
        return client.request("POST", "/tools/analyze-code", req, JsonNode.class);
    }

    /**
     * POST /v1/screenshot(网页截图)。
     *
     * @param req 请求体(url / width / height 等)
     * @return 截图结果(JsonNode,含 imageUrl)
     */
    public JsonNode screenshot(Map<String, Object> req) {
        return client.request("POST", "/screenshot", req, JsonNode.class);
    }
}

File: packages/sdk/dotnet/Modules/ToolsApi.cs
using System.Text.Json;

namespace Ihui.AI;

/// <summary>
/// MCP 工具 / 技能 / 人格 / 代码搜索 / 截图模块。
/// </summary>
/// <remarks>
/// 端点(16 个):
/// <list type="bullet">
///   <item>GET  /v1/tools</item>
///   <item>POST /v1/tools/call</item>
///   <item>GET  /v1/resources</item>
///   <item>GET  /v1/resources/:uri</item>
///   <item>GET  /v1/prompts</item>
///   <item>POST /v1/prompts/invoke</item>
///   <item>GET  /v1/skills</item>
///   <item>GET  /v1/slash-commands</item>
///   <item>POST /v1/slash-commands</item>
///   <item>POST /v1/sampling</item>
///   <item>GET  /v1/personas</item>
///   <item>GET  /v1/personas/:name</item>
///   <item>POST /v1/tools/search-codebase</item>
///   <item>POST /v1/tools/search-web</item>
///   <item>POST /v1/tools/analyze-code</item>
///   <item>POST /v1/screenshot</item>
/// </list>
/// </remarks>
public sealed class ToolsApi
{
    private readonly BaseClient _client;

    internal ToolsApi(BaseClient client) => _client = client;

    /// <summary>GET /v1/tools(MCP 工具列表)。</summary>
    /// <param name="cancellationToken">取消令牌</param>
    /// <returns>工具列表(JsonElement)</returns>
    public Task<JsonElement?> ListAsync(CancellationToken cancellationToken = default)
        => _client.RequestAsync<JsonElement>("GET", "/tools", null, cancellationToken);

    /// <summary>POST /v1/tools/call(调用 MCP 工具)。</summary>
    /// <param name="req">工具调用请求</param>
    /// <param name="cancellationToken">取消令牌</param>
    /// <returns>调用结果(JsonElement)</returns>
    public Task<JsonElement?> CallAsync(ToolCallRequest req, CancellationToken cancellationToken = default)
        => _client.RequestAsync<JsonElement>("POST", "/tools/call", req, cancellationToken);

    /// <summary>GET /v1/resources(MCP 资源列表)。</summary>
    /// <param name="cancellationToken">取消令牌</param>
    /// <returns>资源列表(JsonElement)</returns>
    public Task<JsonElement?> ListResourcesAsync(CancellationToken cancellationToken = default)
        => _client.RequestAsync<JsonElement>("GET", "/resources", null, cancellationToken);

    /// <summary>GET /v1/resources/:uri(资源详情)。</summary>
    /// <param name="uri">资源 URI</param>
    /// <param name="cancellationToken">取消令牌</param>
    /// <returns>资源详情(JsonElement)</returns>
    public Task<JsonElement?> GetResourceAsync(string uri, CancellationToken cancellationToken = default)
        => _client.RequestAsync<JsonElement>("GET", "/resources/" + BaseClient.Encode(uri), null, cancellationToken);

    /// <summary>GET /v1/prompts(MCP 提示词列表)。</summary>
    /// <param name="cancellationToken">取消令牌</param>
    /// <returns>提示词列表(JsonElement)</returns>
    public Task<JsonElement?> ListPromptsAsync(CancellationToken cancellationToken = default)
        => _client.RequestAsync<JsonElement>("GET", "/prompts", null, cancellationToken);

    /// <summary>POST /v1/prompts/invoke(调用提示词)。</summary>
    /// <param name="req">请求体(name / arguments 等)</param>
    /// <param name="cancellationToken">取消令牌</param>
    /// <returns>调用结果(JsonElement)</returns>
    public Task<JsonElement?> InvokePromptAsync(object req, CancellationToken cancellationToken = default)
        => _client.RequestAsync<JsonElement>("POST", "/prompts/invoke", req, cancellationToken);

    /// <summary>GET /v1/skills(技能列表)。</summary>
    /// <param name="cancellationToken">取消令牌</param>
    /// <returns>技能列表(JsonElement)</returns>
    public Task<JsonElement?> ListSkillsAsync(CancellationToken cancellationToken = default)
        => _client.RequestAsync<JsonElement>("GET", "/skills", null, cancellationToken);

    /// <summary>GET /v1/slash-commands(slash 命令列表)。</summary>
    /// <param name="cancellationToken">取消令牌</param>
    /// <returns>命令列表(JsonElement)</returns>
    public Task<JsonElement?> ListSlashCommandsAsync(CancellationToken cancellationToken = default)
        => _client.RequestAsync<JsonElement>("GET", "/slash-commands", null, cancellationToken);

    /// <summary>POST /v1/slash-commands(调用 slash 命令)。</summary>
    /// <param name="req">请求体(command / args)</param>
    /// <param name="cancellationToken">取消令牌</param>
    /// <returns>调用结果(JsonElement)</returns>
    public Task<JsonElement?> InvokeSlashCommandAsync(object req, CancellationToken cancellationToken = default)
        => _client.RequestAsync<JsonElement>("POST", "/slash-commands", req, cancellationToken);

    /// <summary>POST /v1/sampling(模型采样)。</summary>
    /// <param name="req">请求体(messages / model 等)</param>
    /// <param name="cancellationToken">取消令牌</param>
    /// <returns>采样结果(JsonElement)</returns>
    public Task<JsonElement?> SamplingAsync(object req, CancellationToken cancellationToken = default)
        => _client.RequestAsync<JsonElement>("POST", "/sampling", req, cancellationToken);

    /// <summary>GET /v1/personas(人格列表)。</summary>
    /// <param name="cancellationToken">取消令牌</param>
    /// <returns>人格列表(JsonElement)</returns>
    public Task<JsonElement?> ListPersonasAsync(CancellationToken cancellationToken = default)
        => _client.RequestAsync<JsonElement>("GET", "/personas", null, cancellationToken);

    /// <summary>GET /v1/personas/:name(人格详情)。</summary>
    /// <param name="name">人格名称</param>
    /// <param name="cancellationToken">取消令牌</param>
    /// <returns>人格详情(JsonElement)</returns>
    public Task<JsonElement?> GetPersonaAsync(string name, CancellationToken cancellationToken = default)
        => _client.RequestAsync<JsonElement>("GET", "/personas/" + BaseClient.Encode(name), null, cancellationToken);

    /// <summary>POST /v1/tools/search-codebase(代码库搜索)。</summary>
    /// <param name="req">请求体(query / filePattern 等)</param>
    /// <param name="cancellationToken">取消令牌</param>
    /// <returns>搜索结果(JsonElement)</returns>
    public Task<JsonElement?> SearchCodebaseAsync(object req, CancellationToken cancellationToken = default)
        => _client.RequestAsync<JsonElement>("POST", "/tools/search-codebase", req, cancellationToken);

    /// <summary>POST /v1/tools/search-web(网页搜索)。</summary>
    /// <param name="req">请求体(query / count 等)</param>
    /// <param name="cancellationToken">取消令牌</param>
    /// <returns>搜索结果(JsonElement)</returns>
    public Task<JsonElement?> SearchWebAsync(object req, CancellationToken cancellationToken = default)
        => _client.RequestAsync<JsonElement>("POST", "/tools/search-web", req, cancellationToken);

    /// <summary>POST /v1/tools/analyze-code(代码分析)。</summary>
    /// <param name="req">请求体(code / language / analysis 等)</param>
    /// <param name="cancellationToken">取消令牌</param>
    /// <returns>分析结果(JsonElement)</returns>
    public Task<JsonElement?> AnalyzeCodeAsync(object req, CancellationToken cancellationToken = default)
        => _client.RequestAsync<JsonElement>("POST", "/tools/analyze-code", req, cancellationToken);

    /// <summary>POST /v1/screenshot(网页截图)。</summary>
    /// <param name="req">请求体(url / width / height 等)</param>
    /// <param name="cancellationToken">取消令牌</param>
    /// <returns>截图结果(JsonElement,含 imageUrl)</returns>
    public Task<JsonElement?> ScreenshotAsync(object req, CancellationToken cancellationToken = default)
        => _client.RequestAsync<JsonElement>("POST", "/screenshot", req, cancellationToken);
}
